### PR TITLE
Add package name to RustBufferI if it needs it

### PR DESCRIPTION
**Pull request by kegsay**
Fixes https://github.com/NordSecurity/uniffi-bindgen-go/issues/43

### CODE REVIEW

**Comment by arg0d** [comment]
I'm trying to wrap my head around when does this happen. Is there a simple way to repro this inside the tests? I think there are already some tests for "external" types `ext_types_test.go`.

**Comment by kegsay** [comment]
It only happens on async functions when the value being returned is in another package. In that case, both packages have `RustBufferI` types and the compiler complains as the generated code uses both types when it should be choosing one and sticking with it.

There isn't a simple way to reproduce the test due to #45 - I could probably write a test in a completely separate test package and then `go test` that?

**Comment by arg0d** [comment]
I can reproduce #45, but this PR does not seem to do anything for #45, I still get the same error with these changes.

**Comment by kegsay** [comment]
It won't do anything for #45. This PR is fixing #43.

**Comment by kegsay** [comment]
Yup, I have a working regression test now in a new package `uniffi_go_issue43`. The test itself just does:
```
#[uniffi::export]
async fn get_async_external_type() -> uniffi_go_issue45::Record {
    uniffi_go_issue45::Record {
        id: "foo".to_string(),
        tag: "bar".to_string(),
    }
}
```
which requires this PR to produce code that compiles, which produces a `completeFunc` of:
```go
		func(handle *C.void, status *C.RustCallStatus) issue45.RustBufferI {
			// completeFunc
			return C.ffi_uniffi_go_issue43_rust_future_complete_rust_buffer(unsafe.Pointer(handle), status)
		},
```
whereas without #44 it would just be `func(handle *C.void, status *C.RustCallStatus) RustBufferI {` which compile errors on the `liftFunc`: `type func(rb issue45.RustBufferI) issue45.Record of issue45.FfiConverterTypeRecordINSTANCE.Lift does not match inferred type func(RustBufferI) U for func(T) Ucompiler[CannotInferTypeArgs](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#CannotInferTypeArgs)`.

**Comment by kegsay** [comment]
Rebased as #46 landed. It should be pretty clear now!

File: bindgen/src/gen_go/mod.rs [hunk 1]
```diff
@@ -475,6 +475,21 @@ pub mod filters {
         Ok(result)
     }
 
+    // Return the Go package name for the given type, if it has one, followed by a '.'
+    // Returns the empty string if the type has no package name associated with it e.g primitive types.
+    //
+    // for multi-package bindings, it may be required to specify which
+    // RustBufferI should be used when creating completeFunc. Failure to
+    // specify the right package here will result in compilation errors.
+    pub fn maybe_namespace(type_: &impl AsType) -> Result<String, askama::Error> {
```
Comment (arg0d) [inline bindgen/src/gen_go/mod.rs:484]: I would implement this as part of `ffi_type_name` function itself, that would be more robust.
```diff
diff --git a/bindgen/src/gen_go/mod.rs b/bindgen/src/gen_go/mod.rs
index 1c1c638..9781ea7 100644
--- a/bindgen/src/gen_go/mod.rs
+++ b/bindgen/src/gen_go/mod.rs
@@ -465,11 +465,14 @@ pub mod filters {
     }
 
     /// FFI type name to be used to reference cgo types
-    pub fn ffi_type_name<T: Clone + Into<FfiType>>(type_: &T) -> Result<String, askama::Error> {
+    pub fn ffi_type_name(type_: &Type) -> Result<String, askama::Error> {
         let ffi_type: FfiType = type_.clone().into();
         let result = match ffi_type {
             FfiType::RustArcPtr(_) => "unsafe.Pointer".into(),
-            FfiType::RustBuffer(_) => "RustBufferI".into(),
+            FfiType::RustBuffer(_) => match type_ {
+                Type::External { namespace, .. } => format!("{}.RustBufferI", namespace),
+                _ => "RustBufferI".into(),
+            },
             _ => format!("C.{}", oracle().ffi_type_label(&ffi_type)),
         };
         Ok(result)
```

**Comment by arg0d** [comment]
Nice work!

I finally understand why this error only happens with async :sweat_smile: My initial reaction was that this should also happen with regular functions, e.g.
```golang
func GetAsyncExternalType() issue45.Record {
	return issue45.FfiConverterTypeRecordINSTANCE.Lift(rustCall(func(_uniffiStatus *C.RustCallStatus) RustBufferI {
		return C.uniffi_uniffi_go_issue43_fn_func_get_async_external_type(_uniffiStatus)
	}))
}
```
In this example there is also a type mismatch between what `issue45.FfiConverterTypeRecordINSTANCE.Lift` expects, and what `rustCall` returns. But I assume in this case Go's interface duck typing kicks in, and allows an implicit conversion from `issue43.RustBufferI` into `issue45.RustBufferI`.

Contrast that with async code, where `issue43.RustBufferI` and `issue45.RustBufferI` are parameters to a generic function `func uniffiRustCallAsyncWithResult[T any, U any] ..`, and interface duck typing rules do not apply - compiler expects identical types for generic parameters.

**Review by arg0d — CHANGES_REQUESTED** [review]

**Review by arg0d — APPROVED** [review]
Nice work, thanks for contributing!

**Pull request merged**